cdap-docs/developers-manual/source/building-blocks/workflows.rst

```diff
@@ -246,7 +246,53 @@ The `WorkflowToken Java API
 includes methods for getting values for different keys, scopes, and nodes. The same
 key can be added to the workflow by different nodes, and there are methods to return a map of those
 key-value pairs. Convenience methods allow the putting and getting of non-string values
-through the use of the class Value.
+through the use of the class ``Value``.
+
+MapReduce Counters and Workflow Tokens
+--------------------------------------
+When a workflow executes a MapReduce program, MapReduce counters generated by the program are added to the
+workflow token under the *system* scope. Counters can be defined by the MapReduce framework or applications.
+
+The counters defined by the MapReduce framework can be retrieved from the workflow token by using
+a key composed of the counter group name, followed by a ".", followed by the name of the counter.
+For example: to access the number of input records to the map method of the ``PurchaseHistoryBuilder``
+MapReduce program::
+
+  String counterGroupName = "org.apache.hadoop.mapreduce.TaskCounter";
+  String counterName = "MAP_INPUT_RECORDS";
+  String counterKey = counterGroupName + "." + counterName;
+  String counterValue = workflowToken.get(counterKey, "PurchaseHistoryBuilder", WorkflowToken.Scope.SYSTEM);
+
+Applications can define the counters using ``enum`` as::
+
+  // Application defined counters using enum
+  public static enum MY_COUNTER {
+    BY_TYPE,
+    BY_NAME
+  }
+
+  // Access the counters in the MapReduce program
+  context.getCounter(MY_COUNTER.BY_TYPE).increment(1);
+
+These counters can be retrieved from the workflow token by using a key composed of class name of the
+``enum`` as the counter group name, followed by a ".", followed by the ``enum`` member as the name of the counter::
+
+  String counterGroupName = MY_COUNTER.class.getName();
+  String counterName = "BY_TYPE";
+  String counterKey = counterGroupName + "." + counterName;
+  String counterValue = workflowToken.get(counterKey, "PurchaseHistoryBuilder", WorkflowToken.Scope.SYSTEM);
+
+Applications can also define counters by explicitly providing the counter group name and counter name as::
+
+  context.getCounter("MyCounterGroup", "MyCounter").increment(1);
+
+These counters can be retrieved from the workflow token by using a key composed of explicitly provided
+counter group name, followed by a ".", followed by the explicitly provided name of the counter::
+
+  String counterGroupName = "MyCounterGroup";
+  String counterName = "MyCounter";
+  String counterKey = counterGroupName + "." + counterName;
+  String counterValue = workflowToken.get(counterKey, "PurchaseHistoryBuilder", WorkflowToken.Scope.SYSTEM);
```

Comment:
Made sure that the example works. One change I did is to get the counter values as `long`, since that is the way the counters would be accessed.